application/comit_node/src/swap_protocols/rfc003/ethereum/htlc_events.rs

```diff
@@ -0,0 +1,223 @@
+use crate::{
+    ledger_query_service::{EthereumQuery, EventMatcher, QueryEthereum, Topic},
+    swap_protocols::{
+        asset::Asset,
+        ledger::Ethereum,
+        rfc003::{
+            self,
+            ethereum::{extract_secret::extract_secret, REDEEM_LOG_MSG, REFUND_LOG_MSG},
+            events::{Deployed, Funded, HtlcEvents, RedeemedOrRefunded},
+            state_machine::HtlcParams,
+            FundTransaction, RedeemTransaction, RefundTransaction,
+        },
+    },
+};
+use ethereum_support::{
+    web3::types::Address, CalculateContractAddress, Erc20Token, EtherQuantity, Transaction,
+};
+use futures::{
+    future::{self, Either},
+    Future,
+};
+use std::sync::Arc;
+
+impl HtlcEvents<Ethereum, EtherQuantity> for Arc<dyn QueryEthereum + Send + Sync + 'static> {
+    fn htlc_deployed(
+        &self,
+        htlc_params: HtlcParams<Ethereum, EtherQuantity>,
+    ) -> Box<Deployed<Ethereum>> {
+        let query_ethereum = Arc::clone(&self);
+        let query = query_ethereum
+            .create(EthereumQuery::Transaction {
+                from_address: None,
+                to_address: None,
+                is_contract_creation: Some(true),
+                transaction_data: Some(htlc_params.bytecode()),
+                transaction_data_length: None,
+            })
+            .map_err(rfc003::Error::LedgerQueryService);
+
+        let transaction = query.and_then(move |query_id| {
+            query_ethereum
+                .transaction_first_result(&query_id)
+                .map_err(rfc003::Error::LedgerQueryService)
+        });
+
+        let htlc_location = transaction.and_then(move |tx| {
+            let actual_value = EtherQuantity::from_wei(tx.value);
+            let required_value = htlc_params.asset;
+
+            if actual_value < required_value {
```

Comment:
:x: I really don't like that functionality like this is not abstracted away by a trait but implemented next to some other boilerplate code that glues stuff together.

Let's discuss this.

Reply:
I moved the responsibility fo checking the asset into the state machine itself -- i.e. the state machine shouldn't transition to Funded unless the asset that the HTLC was funded with is of greater or equal value to the asset that was agreed upon.

Reply:
That is a brilliant idea!